packages/hash/api/src/graphql/resolvers/pages/updatePageContents.ts

```diff
@@ -1,70 +1,239 @@
 import { ApolloError, UserInputError } from "apollo-server-errors";
-
+import { produce } from "immer";
 import {
-  Resolver,
+  Block,
+  Entity,
+  EntityType,
+  Page,
+  UnresolvedGQLEntity,
+} from "../../../model";
+import { exactlyOne } from "../../../util";
+import {
+  EntityDefinition,
   MutationUpdatePageContentsArgs,
-  UpdatePageAction,
-  UpdateEntity,
+  Resolver,
   SwapBlockData,
+  UpdateEntity,
+  UpdatePageAction,
+  UpdatePageContentsResult,
 } from "../../apiTypes.gen";
-import { Block, Entity, Page, UnresolvedGQLEntity } from "../../../model";
 import { LoggedInGraphQLContext } from "../../context";
-import { exactlyOne } from "../../../util";
 
 const validateActionsInput = (actions: UpdatePageAction[]) => {
   for (const [i, action] of actions.entries()) {
     if (
       !exactlyOne(
-        action.insertNewBlock,
+        action.insertBlock,
         action.moveBlock,
         action.removeBlock,
         action.updateEntity,
         action.swapBlockData,
+        action.createEntity,
+        action.createEntityType,
       )
     ) {
       throw new UserInputError(
-        `at action ${i}: exactly one of insertNewBlock, moveBlock, removeBlock or updateEntity must be specified`,
+        `at action ${i}: exactly one of insertBlock, moveBlock, removeBlock or updateEntity must be specified`,
       );
     }
   }
 };
 
+const isPlaceholder = (id: unknown): id is string =>
+  typeof id === "string" && id.startsWith("placeholder-");
+
+// @todo these actions need to be processed in order to ensure placeholders
+// work as expected
 export const updatePageContents: Resolver<
-  Promise<UnresolvedGQLEntity>,
+  Promise<
+    {
+      page: UnresolvedGQLEntity;
+    } & Omit<UpdatePageContentsResult, "page">
+  >,
   {},
   LoggedInGraphQLContext,
   MutationUpdatePageContentsArgs
-> = async (_, { accountId, entityId, actions }, { dataSources, user }) => {
+> = async (
+  _,
+  { accountId, entityId: pageEntityId, actions },
+  { dataSources, user },
+) => {
   validateActionsInput(actions);
 
+  /**
+   * Some actions allow you to put in placeholder entityIds which refer to
+   * previous entities created during this mutation. These must always start
+   * with "placeholder-".
+   */
+  const placeholderResults = new Map<string, string>();
+
+  const replacePlaceholder = (placeholder: string) => {
```

Comment:
Agree... It's a bit difficult to know what `replacePlaceholder` does since it doesn't exactly replace the placeholderId.

I think `getEntityIdFromPlaceholderId` is a bit more descriptive. It's also similar to the existing `getDraftEntityFromEntityId` method we have. I guess the only difference is it returns the argument passed in if it's not a placeholderId

Reply:
Gone for `getEntityIdFromPossiblePlaceholderId`